### DISCUSSION TITLE
Using dynamic import with loading function and useSWR in the component that is loaded

### DISCUSSION THREAD

**Discussion by danechitoaie** [vercel/next.js#37867] [Help]
Hi,

In my `index.tsx` I have
```typescript
import type { NextPage } from "next";
import dynamic from "next/dynamic";

const Challenge = dynamic(() => import("../components/Challenge"), {
    ssr: false,
    loading: () => <div>loading...</div>,
});

const Index: NextPage = () => {
    return <Challenge />;
};

export default Index;
```

and in `../components/Challenge.tsx` I have
```typescript
export default function Challenge() {
    const { data, error } = useSWR("/api/data", (...args) =>
        fetch(...args).then((res) => res.json())
    );

    return (
        <>
            ...
        </>
    );
}
```

Any idea how this works? Will `<div>loading...</div>` be displayed only while `../components/Challenge.tsx` is loaded or it will also wait for the `useSWR` function to load the data?

**Comment by icyJoseph** [comment]
As far as I know, this will first, load the component JavaScript, and while it does that, it shows `loading...`, then once the component is downloaded and parsed, as part of its React life cycle, the custom SWR hook is executed.

This is what some peopler refer as waterfalls, because one async task, triggers yet another async task, in order to see something in the screen.

Because you are in Next.js, you have control over more things that in a regular CSR, so you could do this:

```tsx
<link rel="preload" href="/api/data" as="fetch" crossorigin="anonymous" />
```

Add this between Head tags to this page, that way the browser will go out and fetch data for that resource ASAP (even while it is downloading other resources), then when the JavaScript for the Challenge component is done, SWR will quickly access the already fetched response. Taken directly from: https://swr.vercel.app/docs/prefetching

**Comment by danechitoaie** [comment]
Is there any way to wait for both to load? I want to have a full page loading animation. if I once have a loader with `loading: () => <div>loading...</div>,` and then again inside `../components/Challenge.tsx`  by checking `if (!data)` the I kind of end up with a loader inside a loader.
